julius-4.2.2/man/julius.1
.\"     Title: 
    julius
  
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.71.0 <http://docbook.sf.net/>
.\"      Date: 02/11/2009
.\"    Manual: 
.\"    Source: 
.\"
.TH "JULIUS" "1" "02/11/2009" "" ""
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"

    julius
   \- open source multi\-purpose LVCSR engine
.SH "SYNOPSIS"
.HP 7
\fBjulius\fR [\-C\ \fIjconffile\fR] [\fIoptions\fR...]
.SH "DESCRIPTION"
.PP

\fBjulius\fR
is a high\-performance, multi\-purpose, open\-source speech recognition engine for researchers and developers. It is capable of performing almost real\-time recognition of continuous speech with over 60k\-word 3\-gram language model and triphone HMM model, on most current PCs.
\fBjulius\fR
can perform recognition on audio files, live microphone input, network input and feature parameter files.
.PP
The core recognition module is implemented as C library called "JuliusLib". It can also be extended by plug\-in facility.
.SS "Supported Models"
.PP
\fBjulius\fR
needs a language model and an acoustic model to run as a speech recognizer.
\fBjulius\fR
supports the following models.
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBAcoustic model\fR
.RS
.PP
Sub\-word HMM (Hidden Markov Model) in
HTK
ascii format are supported. Phoneme models (monophone), context dependent phoneme models (triphone), tied\-mixture and phonetic tied\-mixture models of any unit can be used. When using context dependent models, inter\-word context dependency is also handled. Multi\-stream feature and MSD\-HMM is also supported. You can further use a tool
\fBmkbinhmm\fR
to convert the ascii HMM file to a compact binary format for faster loading.
.PP
Note that
\fBjulius\fR
itself can only extract MFCC features from speech data. If you use acoustic HMM trained for other feature, you should give the input in
HTK
parameter file of the same feature type.
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBLanguage model: word N\-gram\fR
.RS
.PP
Word N\-gram language model, up to 10\-gram, is supported. Julius uses different N\-gram for each pass: left\-to\-right 2\-gram on 1st pass, and right\-to\-left N\-gram on 2nd pass. It is recommended to use both LR 2\-gram and RL N\-gram for Julius. However, you can use only single LR N\-gram or RL N\-gram. In such case, approximated LR 2\-gram computed from the given N\-gram will be applied at the first pass.
.PP
The Standard ARPA format is supported. In addition, a binary format is also supported for efficiency. The tool
\fBmkbingram\fR(1)
can convert ARPA format N\-gram to binary format.
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBLanguage model: grammar\fR
.RS
.PP
The grammar format is an original one, and tools to create a recognirion grammar are included in the distribution. A grammar consists of two files: one is a 'grammar' file that describes sentence structures in a BNF style, using word 'category' name as terminate symbols. Another is a 'voca' file that defines words with its pronunciations (i.e. phoneme sequences) for each category. They should be converted by
\fBmkdfa.pl\fR(1)
to a deterministic finite automaton file (.dfa) and a dictionary file (.dict), respectively. You can also use multiple grammars.
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBLanguage model: isolated word\fR
.RS
.PP
You can perform isolated word recognition using only word dictionary. With this model type, Julius will perform rapid one pass recognition with static context handling. Silence models will be added at both head and tail of each word. You can also use multiple dictionaries in a process.
.RE
.SS "Search Algorithm"
.PP
Recognition algorithm of
\fBjulius\fR
is based on a two\-pass strategy. Word 2\-gram and reverse word 3\-gram is used on the respective passes. The entire input is processed on the first pass, and again the final searching process is performed again for the input, using the result of the first pass to narrow the search space. Specifically, the recognition algorithm is based on a tree\-trellis heuristic search combined with left\-to\-right frame\-synchronous beam search and right\-to\-left stack decoding search.
.PP
When using context dependent phones (triphones), interword contexts are taken into consideration. For tied\-mixture and phonetic tied\-mixture models, high\-speed acoustic likelihood calculation is possible using gaussian pruning.
.PP
For more details, see the related documents.
.SH "OPTIONS"
.PP
These options specify the models, system behaviors and various search parameters to Julius. These option can be set at the command line, but it is recommended that you write them in a text file as a "jconf file", and specify it by "\-C" option.
.PP
Applications incorporating JuliusLib also use these options to set the parameters of core recognition engine. For example, a jconf file can be loaded to the enine by calling
\fBj_config_load_file_new()\fR
with the jconf file name as argument.
.PP
Please note that relative paths in a jconf file should be relative to the jconf file itself, not the current working directory.
.PP
Below are the details of all options, gathered by group.
.SS "Julius application option"
.PP
These are application options of Julius, outside of JuliusLib. It contains parameters and switches for result output, character set conversion, log level, and module mode options. These option are specific to Julius, and cannot be used at applications using JuliusLib other than Julius.
.PP
\fB \-outfile \fR
.RS 3n
On file input, this option write the recognition result of each file to a separate file. The output file of an input file will be the same name but the suffix will be changed to ".out". (rev.4.0)
.RE
.PP
\fB \-separatescore \fR
.RS 3n
Output the language and acoustic scores separately.
.RE
.PP
\fB \-callbackdebug \fR
.RS 3n
Print the callback names at each call for debug. (rev.4.0)
.RE
.PP
\fB \-charconv \fR \fIfrom\fR \fIto\fR
.RS 3n
Print with character set conversion.
\fIfrom\fR
is the source character set used in the language model, and
\fIto\fR
is the target character set you want to get.
.sp
On Linux, the arguments should be a code name. You can obtain the list of available code names by invoking the command "iconv \-\-list". On Windows, the arguments should be a code name or codepage number. Code name should be one of "ansi", "mac", "oem", "utf\-7", "utf\-8", "sjis", "euc". Or you can specify any codepage number supported at your environment.
.RE
.PP
\fB \-nocharconv \fR
.RS 3n
Disable character conversion.
.RE
.PP
\fB \-module \fR [port]
.RS 3n
Run Julius on "Server Module Mode". After startup, Julius waits for tcp/ip connection from client. Once connection is established, Julius start communication with the client to process incoming commands from the client, or to output recognition results, input trigger information and other system status to the client. The default port number is 10500.
.RE
.PP
\fB \-record \fR \fIdir\fR
.RS 3n
Auto\-save all input speech data into the specified directory. Each segmented inputs are recorded each by one. The file name of the recorded data is generated from system time when the input ends, in a style of
YYYY.MMDD.HHMMSS.wav. File format is 16bit monoral
WAV. Invalid for mfcfile input.
.sp
With input rejection by
\fB\-rejectshort\fR, the rejected input will also be recorded even if they are rejected.
.RE
.PP
\fB \-logfile \fR \fIfile\fR
.RS 3n
Save all log output to a file instead of standard output. (Rev.4.0)
.RE
.PP
\fB \-nolog \fR
.RS 3n
Disable all log output. (Rev.4.0)
.RE
.PP
\fB \-help \fR
.RS 3n
Output help message and exit.
.RE
.SS "Global options"
.PP
These are model\-/search\-dependent options relating audio input, sound detection, GMM, decoding algorithm, plugin facility, and others. Global options should be placed before any instance declaration (\fB\-AM\fR,
\fB\-LM\fR, or
\fB\-SR\fR), or just after "\fB\-GLOBAL\fR" option.
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBAudio input\fR
.RS
.PP
\fB \-input \fR {mic|rawfile|mfcfile|adinnet|stdin|netaudio|alsa|oss|esd}
.RS 3n
Choose speech input source. Specify 'file' or 'rawfile' for waveform file, 'htkparam' or 'mfcfile' for
HTK
parameter file. On file input, users will be prompted to enter the file name from stdin, or you can use
\fB\-filelist\fR
option to specify list of files to process.
.sp
\'mic' is to get audio input from a default live microphone device, and 'adinnet' means receiving waveform data via tcpip network from an adinnet client. 'netaudio' is from DatLink/NetAudio input, and 'stdin' means data input from standard input.
.sp
For waveform file input, only
WAV
(no compression) and
RAW
(noheader, 16bit, big endian) are supported by default. Other format can be read when compiled with
libsnd
library. To see what format is actually supported, see the help message using option
\fB\-help\fR. For stdin input, only WAV and RAW is supported. (default: mfcfile)
.sp
At Linux, you can choose API at run time by specifying
alsa,
oss
and
esd.
.RE
.PP
\fB \-chunk_size \fR \fIsamples\fR
.RS 3n
Audio fragment size in number of samples. (default: 1000)
.RE
.PP
\fB \-filelist \fR \fIfilename\fR
.RS 3n
(With
\fB\-input rawfile|mfcfile\fR) perform recognition on all files listed in the file. The file should contain input file per line. Engine will end when all of the files are processed.
.RE
.PP
\fB \-notypecheck \fR
.RS 3n
By default, Julius checks the input parameter type whether it matches the AM or not. This option will disable the check and force engine to use the input vector as is.
.RE
.PP
\fB \-48 \fR
.RS 3n
Record input with 48kHz sampling, and down\-sample it to 16kHz on\-the\-fly. This option is valid for 16kHz model only. The down\-sampling routine was ported from
sptk. (Rev. 4.0)
.RE
.PP
\fB \-NA \fR \fIdevicename\fR
.RS 3n
Host name for DatLink server input (\fB\-input netaudio\fR).
.RE
.PP
\fB \-adport \fR \fIport_number\fR
.RS 3n
With
\fB\-input adinnet\fR, specify adinnet port number to listen. (default: 5530)
.RE
.PP
\fB \-nostrip \fR
.RS 3n
Julius by default removes successive zero samples in input speech data. This option inhibits the removal.
.RE
.PP
\fB \-zmean \fR, \fB \-nozmean \fR
.RS 3n
This option enables/disables DC offset removal of input waveform. Offset will be estimated from the whole input. For microphone / network input, zero mean of the first 48000 samples (3 seconds in 16kHz sampling) will be used for the estimation. (default: disabled)
.sp
This option uses static offset for the channel. See also
\fB\-zmeansource\fR
for frame\-wise offset removal.
.RE
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBSpeech detection by level and zero\-cross\fR
.RS
.PP
\fB \-cutsilence \fR, \fB \-nocutsilence \fR
.RS 3n
Turn on / off the speech detection by level and zero\-cross. Default is on for mic / adinnet input, and off for files.
.RE
.PP
\fB \-lv \fR \fIthres\fR
.RS 3n
Level threshold for speech input detection. Values should be in range from 0 to 32767. (default: 2000)
.RE
.PP
\fB \-zc \fR \fIthres\fR
.RS 3n
Zero crossing threshold per second. Only input that goes over the level threshold (\fB\-lv\fR) will be counted. (default: 60)
.RE
.PP
\fB \-headmargin \fR \fImsec\fR
.RS 3n
Silence margin at the start of speech segment in milliseconds. (default: 300)
.RE
.PP
\fB \-tailmargin \fR \fImsec\fR
.RS 3n
Silence margin at the end of speech segment in milliseconds. (default: 400)
.RE
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBInput rejection\fR
.RS
.PP
Two simple front\-end input rejection methods are implemented, based on input length and average power of detected segment. The rejection by average power is experimental, and can be enabled by
\-\-enable\-power\-reject
on compilation. Valid for MFCC feature with power coefficient and real\-time input only.
.PP
For GMM\-based input rejection see the GMM section below.
.PP
\fB \-rejectshort \fR \fImsec\fR
.RS 3n
Reject input shorter than specified milliseconds. Search will be terminated and no result will be output.
.RE
.PP
\fB \-powerthres \fR \fIthres\fR
.RS 3n
Reject the inputted segment by its average energy. If the average energy of the last recognized input is below the threshold, Julius will reject the input. (Rev.4.0)
.sp
This option is valid when
\-\-enable\-power\-reject
is specified at compilation time.
.RE
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBGaussian mixture model / GMM\-VAD\fR
.RS
.PP
GMM will be used for input rejection by accumulated score, or for front\-end GMM\-based VAD when
\-\-enable\-gmm\-vad
is specified.
.PP
NOTE: You should also set the proper MFCC parameters required for the GMM, specifying the acoustic parameters described in AM section
\fB\-AM_GMM\fR.
.PP
When GMM\-based VAD is enabled, the voice activity score will be calculated at each frame as front\-end processing. The value will be computed as \\[ \\max_{m \\in M_v} p(x|m) \- \\max_{m \\in M_n} p(x|m) \\] where $M_v$ is a set of voice GMM, and $M_n$ is a set of noise GMM whose names should be specified by
\fB\-gmmreject\fR. The activity score will be then averaged for the last N frames, where N is specified by
\fB\-gmmmargin\fR. Julius updates the averaged activity score at each frame, and detect speech up\-trigger when the value gets higher than a value specified by
\fB\-gmmup\fR, and detecgt down\-trigger when it gets lower than a value of
\fB\-gmmdown\fR.
.PP
\fB \-gmm \fR \fIhmmdefs_file\fR
.RS 3n
GMM definition file in
HTK
format. If specified, GMM\-based input verification will be performed concurrently with the 1st pass, and you can reject the input according to the result as specified by
\fB\-gmmreject\fR. The GMM should be defined as one\-state HMMs.
.RE
.PP
\fB \-gmmnum \fR \fInumber\fR
.RS 3n
Number of Gaussian components to be computed per frame on GMM calculation. Only the N\-best Gaussians will be computed for rapid calculation. The default is 10 and specifying smaller value will speed up GMM calculation, but too small value (1 or 2) may cause degradation of identification performance.
.RE
.PP
\fB \-gmmreject \fR \fIstring\fR
.RS 3n
Comma\-separated list of GMM names to be rejected as invalid input. When recognition, the log likelihoods of GMMs accumulated for the entire input will be computed concurrently with the 1st pass. If the GMM name of the maximum score is within this string, the 2nd pass will not be executed and the input will be rejected.
.RE
.PP
\fB \-gmmmargin \fR \fIframes\fR
.RS 3n
(GMM_VAD) Head margin in frames. When a speech trigger detected by GMM, recognition will start from current frame minus this value. (Rev.4.0)
.sp
This option will be valid only if compiled with
\-\-enable\-gmm\-vad.
.RE
.PP
\fB \-gmmup \fR \fIvalue\fR
.RS 3n
(GMM_VAD) Up trigger threshold of voice activity score. (Rev.4.1)
.sp
This option will be valid only if compiled with
\-\-enable\-gmm\-vad.
.RE
.PP
\fB \-gmmdown \fR \fIvalue\fR
.RS 3n
(GMM_VAD) Down trigger threshold of voice activity score. (Rev.4.1)
.sp
This option will be valid only if compiled with
\-\-enable\-gmm\-vad.
.RE
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBDecoding option\fR
.RS
.PP
Real\-time processing means concurrent processing of MFCC computation 1st pass decoding. By default, real\-time processing on the pass is on for microphone / adinnet / netaudio input, and for others.
.PP
\fB \-realtime \fR, \fB \-norealtime \fR
.RS 3n
Explicitly switch on / off real\-time (pipe\-line) processing on the first pass. The default is off for file input, and on for microphone, adinnet and NetAudio input. This option relates to the way CMN and energy normalization is performed: if off, they will be done using average features of whole input. If on, MAP\-CMN and energy normalization to do real\-time processing.
.RE
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBMisc. options\fR
.RS
.PP
\fB \-C \fR \fIjconffile\fR
.RS 3n
Load a jconf file at here. The content of the jconffile will be expanded at this point.
.RE
.PP
\fB \-version \fR
.RS 3n
Print version information to standard error, and exit.
.RE
.PP
\fB \-setting \fR
.RS 3n
Print engine setting information to standard error, and exit.
.RE
.PP
\fB \-quiet \fR
.RS 3n
Output less log. For result, only the best word sequence will be printed.
.RE
.PP
\fB \-debug \fR
.RS 3n
(For debug) output enormous internal message and debug information to log.
.RE
.PP
\fB \-check \fR {wchmm|trellis|triphone}
.RS 3n
For debug, enter interactive check mode.
.RE
.PP
\fB \-plugindir \fR \fIdirlist\fR
.RS 3n
Specify directory to load plugin. If several direcotries exist, specify them by colon\-separated list.
.RE
.RE
.SS "Instance declaration for multi decoding"
.PP
The following arguments will create a new configuration set with default parameters, and switch current set to it. Jconf parameters specified after the option will be set into the current set.
.PP
To do multi\-model decoding, these argument should be specified at the first of each model / search instances with different names. Any options before the first instance definition will be IGNORED.
.PP
When no instance definition is found (as older version of Julius), all the options are assigned to a default instance named
_default.
.PP
Please note that decoding with a single LM and multiple AMs is not fully supported. For example, you may want to construct the jconf file as following.
.sp .RS 3n .nf \-AM am_1 \-AM am_2 \-LM lm (LM spec..) \-SR search1 am_1 lm \-SR search2 am_2 lm .fi .RE
This type of model sharing is not supported yet, since some part of LM processing depends on the assigned AM. Instead, you can get the same result by defining the same LMs for each AM, like this:
.sp .RS 3n .nf \-AM am_1 \-AM am_2 \-LM lm_1 (LM spec..) \-LM lm_2 (same LM spec..) \-SR search1 am_1 lm_1 \-SR search2 am_2 lm_2 .fi .RE
.PP
\fB \-AM \fR \fIname\fR
.RS 3n
Create a new AM configuration set, and switch current to the new one. You should give a unique name. (Rev.4.0)
.RE
.PP
\fB \-LM \fR \fIname\fR
.RS 3n
Create a new LM configuration set, and switch current to the new one. You should give a unique name. (Rev.4.0)
.RE
.PP
\fB \-SR \fR \fIname\fR \fIam_name\fR \fIlm_name\fR
.RS 3n
Create a new search configuration set, and switch current to the new one. The specified AM and LM will be assigned to it. The
\fIam_name\fR
and
\fIlm_name\fR
can be either name or ID number. You should give a unique name. (Rev.4.0)
.RE
.PP
\fB \-AM_GMM \fR
.RS 3n
When using GMM for front\-end processing, you can specify GMM\-specific acoustic parameters after this option. If you does not specify
\fB\-AM_GMM\fR
with GMM, the GMM will share the same parameter vector as the last AM. The current AM will be switched to the GMM one, so be careful not to confuse with normal AM configurations. (Rev.4.0)
.RE
.PP
\fB \-GLOBAL \fR
.RS 3n
Start a global section. The global options should be placed before any instance declaration, or after this option on multiple model recognition. This can be used multiple times. (Rev.4.1)
.RE
.PP
\fB \-nosectioncheck \fR, \fB \-sectioncheck \fR
.RS 3n
Disable / enable option location check in multi\-model decoding. When enabled, the options between instance declaration is treated as "sections" and only the belonging option types can be written. For example, when an option
\fB\-AM\fR
is specified, only the AM related option can be placed after the option until other declaration is found. Also, global options should be placed at top, before any instance declarataion. This is enabled by default. (Rev.4.1)
.RE
.SS "Language model (\fB\-LM\fR)"
.PP
This group contains options for model definition of each language model type. When using multiple LM, one instance can have only one LM.
.PP
Only one type of LM can be specified for a LM configuration. If you want to use multi model, you should define them one as a new LM.
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBN\-gram\fR
.RS
.PP
\fB \-d \fR \fIbingram_file\fR
.RS 3n
Use binary format N\-gram. An ARPA N\-gram file can be converted to Julius binary format by
mkbingram.
.RE
.PP
\fB \-nlr \fR \fIarpa_ngram_file\fR
.RS 3n
A forward, left\-to\-right N\-gram language model in standard ARPA format. When both a forward N\-gram and backward N\-gram are specified, Julius uses this forward 2\-gram for the 1st pass, and the backward N\-gram for the 2nd pass.
.sp
Since ARPA file often gets huge and requires a lot of time to load, it may be better to convert the ARPA file to Julius binary format by
mkbingram. Note that if both forward and backward N\-gram is used for recognition, they together will be converted to a single binary.
.sp
When only a forward N\-gram is specified by this option and no backward N\-gram specified by
\fB\-nrl\fR, Julius performs recognition with only the forward N\-gram. The 1st pass will use the 2\-gram entry in the given N\-gram, and The 2nd pass will use the given N\-gram, with converting forward probabilities to backward probabilities by Bayes rule. (Rev.4.0)
.RE
.PP
\fB \-nrl \fR \fIarpa_ngram_file\fR
.RS 3n
A backward, right\-to\-left N\-gram language model in standard ARPA format. When both a forward N\-gram and backward N\-gram are specified, Julius uses the forward 2\-gram for the 1st pass, and this backward N\-gram for the 2nd pass.
.sp
Since ARPA file often gets huge and requires a lot of time to load, it may be better to convert the ARPA file to Julius binary format by
mkbingram. Note that if both forward and backward N\-gram is used for recognition, they together will be converted to a single binary.
.sp
When only a backward N\-gram is specified by this option and no forward N\-gram specified by
\fB\-nlr\fR, Julius performs recognition with only the backward N\-gram. The 1st pass will use the forward 2\-gram probability computed from the backward 2\-gram using Bayes rule. The 2nd pass fully use the given backward N\-gram. (Rev.4.0)
.RE
.PP
\fB \-v \fR \fIdict_file\fR
.RS 3n
Word dictionary file.
.RE
.PP
\fB \-silhead \fR \fIword_string\fR \fB \-siltail \fR \fIword_string\fR
.RS 3n
Silence word defined in the dictionary, for silences at the beginning of sentence and end of sentence. (default: "<s>", "</s>")
.RE
.PP
\fB \-mapunk \fR \fIword_string\fR
.RS 3n
Specify unknown word. Default is "<unk>" or "<UNK>". This will be used to assign word probability on unknown words, i.e. words in dictionary that are not in N\-gram vocabulary.
.RE
.PP
\fB \-iwspword \fR
.RS 3n
Add a word entry to the dictionary that should correspond to inter\-word pauses. This may improve recognition accuracy in some language model that has no explicit inter\-word pause modeling. The word entry to be added can be changed by
\fB\-iwspentry\fR.
.RE
.PP
\fB \-iwspentry \fR \fIword_entry_string\fR
.RS 3n
Specify the word entry that will be added by
\fB\-iwspword\fR. (default: "<UNK> [sp] sp sp")
.RE
.PP
\fB \-sepnum \fR \fInumber\fR
.RS 3n
Number of high frequency words to be isolated from the lexicon tree, to ease approximation error that may be caused by the one\-best approximation on 1st pass. (default: 150)
.RE
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBGrammar\fR
.RS
.PP
Multiple grammars can be specified by repeating
\fB\-gram\fR
and
\fB\-gramlist\fR. Note that this is unusual behavior from other options (in normal Julius option, last one will override previous ones). You can use
\fB\-nogram\fR
to reset the grammars already specified before the point.
.PP
\fB \-gram \fR gramprefix1[,gramprefix2[,gramprefix3,...]]
.RS 3n
Comma\-separated list of grammars to be used. the argument should be a prefix of a grammar, i.e. if you have
\fIfoo.dfa\fR
and
\fIfoo.dict\fR, you should specify them with a single argument
foo. Multiple grammars can be specified at a time as a comma\-separated list.
.RE
.PP
\fB \-gramlist \fR \fIlist_file\fR
.RS 3n
Specify a grammar list file that contains list of grammars to be used. The list file should contain the prefixes of grammars, each per line. A relative path in the list file will be treated as relative to the file, not the current path or configuration file.
.RE
.PP
\fB \-dfa \fR \fIdfa_file\fR \fB \-v \fR \fIdict_file\fR
.RS 3n
An old way of specifying grammar files separately. This is bogus, and should not be used any more.
.RE
.PP
\fB \-nogram \fR
.RS 3n
Remove the current list of grammars already specified by
\fB\-gram\fR,
\fB\-gramlist\fR,
\fB\-dfa\fR
and
\fB\-v\fR.
.RE
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
\fBIsolated word\fR
.RS
.PP
Dictionary can be specified by using
\fB\-w\fR
and
\fB\-wlist\fR. When you specify multiple times, all of them will be read at startup. You can use
\fB\-nogram\fR
to reset the already specified dictionaries at that point.
.PP
\fB \-w \fR \fIdict_file\fR
.RS 3n
Word dictionary for isolated word recognition. File format is the same as other LM. (Rev.4.0)
.RE
.PP
\fB \-wlist \fR \fIlist_file\fR
.RS 3n
Specify a dictionary list file that contains list of dictionaries to be used. The list file should contain the file name of dictionaries, each per line. A relative path in the list file will be treated as relative to the list file, not the current path or configuration file. (Rev.4.0)
.RE
.PP
\fB \-nogram \fR
.RS 3n
Remove the current list of dictionaries already specified by
\fB\-w\fR
and
\fB\-wlist\fR.
.RE
.PP

src/julius.cpp
#include <iostream>
#include <memory>
#include <mutex>
#include "julius.hpp"

using namespace node;
using namespace v8;

/* ------------------------------------------------------------------------- */
//  グローバル変数 / データやりとり用 構造体
/* ------------------------------------------------------------------------- */
// EventEmitter.emit 用のシンボル
static Persistent<String> emit_symbol;

//! Julius の別スレッドからのアクセス用
static uv_mutex_t m;

// Emit でやりとりするデータ
struct EmitData
{
	Julius *_this;
	std::string msg;
	std::string result;
};

/* ------------------------------------------------------------------------- */
//  Static メンバ
/* ------------------------------------------------------------------------- */
void Julius::Init(Handle<Object>& target)
{
	auto clazz = FunctionTemplate::New(Julius::New);
	clazz->SetClassName( String::NewSymbol("Julius") );
	clazz->InstanceTemplate()->SetInternalFieldCount(1);
	clazz->PrototypeTemplate()->Set(
		String::NewSymbol("start"),
		FunctionTemplate::New(Julius::Start)->GetFunction()
	);
	clazz->PrototypeTemplate()->Set(
		String::NewSymbol("reload"),
		FunctionTemplate::New(Julius::Reload)->GetFunction()
	);
	clazz->PrototypeTemplate()->Set(
		String::NewSymbol("stop"),
		FunctionTemplate::New(Julius::Stop)->GetFunction()
	);
	target->Set( String::NewSymbol("Julius"), clazz->GetFunction() );
}


Handle<v8::Value> Julius::New(const Arguments& args)
{
	HandleScope scope;

	String::Utf8Value jconf_path(args[0]);
	std::cout << *jconf_path << std::endl;
	auto _this = new Julius(*jconf_path);
	_this->Wrap( args.This() );

	return scope.Close( args.This() );
}


Handle<v8::Value> Julius::Start(const Arguments& args)
{
	HandleScope scope;
	auto _this = ObjectWrap::Unwrap<Julius>( args.This() );

	switch (j_open_stream(_this->recog_, nullptr)) {
		case  0: break; // success
		case -1: _this->Emit("error", "Error in input stream");        return scope.Close( Boolean::New(false) );
		case -2: _this->Emit("error", "Failed to begin input stream"); return scope.Close( Boolean::New(false) );
	}

	auto req  = new uv_work_t;
	req->data = _this;

	_this->Ref(); // v8 に GC されないようにする
	uv_queue_work(
		uv_default_loop(),
		req,
		[](uv_work_t* req) {
			std::cout << "Start" << std::endl;
			auto _this = static_cast<Julius*>(req->data);
			_this->Emit("start");
			j_recognize_stream(_this->recog_);
		},
		[](uv_work_t* req) {
			auto _this = static_cast<Julius*>(req->data);
			_this->Unref(); // v8 に GC を許可する
		}
	);

	return scope.Close( Boolean::New(true) );
};


Handle<v8::Value> Julius::Stop(const Arguments& args)
{
	HandleScope scope;

	auto _this = ObjectWrap::Unwrap<Julius>( args.This() );
	j_close_stream(_this->recog_);
	_this->Emit("pause");

	return scope.Close( Undefined() );
};

Handle<v8::Value> Julius::Reload(const Arguments& args)
{
	HandleScope scope;

	String::Utf8Value jconf_path(args[0]);
	auto _this = ObjectWrap::Unwrap<Julius>( args.This() );
	j_close_stream(_this->recog_);
	_this->init(*jconf_path);
	_this->Emit("reload");

	return scope.Close( Undefined() );
};

void Julius::Emit(const std::string& msg, const std::string& result)
{
	// やり取りするデータを作詞
	auto req     = new uv_work_t;
	auto data    = new EmitData;
	data->_this  = this;
	data->msg    = msg;
	data->result = result;
	req->data    = data;

	// 次回のループの最初にメインスレッドからコールバックを実行
	uv_queue_work(
		uv_default_loop(),
		req,
		[](uv_work_t* req) {},
		[](uv_work_t* req) {
			uv_mutex_lock(&m);
			HandleScope scope;

			auto data = static_cast<EmitData*>(req->data);
			auto emit_v = data->_this->handle_->Get(emit_symbol);
			assert( emit_v->IsFunction() );
			auto emit = emit_v.As<Function>();

			TryCatch tc;
			Handle<v8::Value> argv[] = { String::New(data->msg.c_str()), String::New(data->result.c_str()) };
			emit->Call(data->_this->handle_, 2, argv);
			if ( tc.HasCaught() ) {
				FatalException(tc);
			}

			uv_mutex_unlock(&m);
		}
	);
}

/* ------------------------------------------------------------------------- */
//  non-Static メンバ
/* ------------------------------------------------------------------------- */
Julius::Julius(const std::string& jconf_path)
: recog_(nullptr)
{
	init(jconf_path);
}


Julius::~Julius()
{
	free_recog();
}


void Julius::init(const std::string& jconf_path)
{
	make_recog(jconf_path);

	// デバッグメッセージを消す
	j_disable_debug_message();
	j_disable_verbose_message();
}


void Julius::make_recog(const std::string& jconf_path)
{
	// 古い recog がある場合は破棄
	free_recog();

	// jconf を読み込む
	std::vector<char> path( jconf_path.begin(), jconf_path.end() );
	path.push_back('\0');
	auto jconf = j_config_load_file_new( &path[0] );
	if (jconf == nullptr) {
		Emit("error", "Error@j_config_load_file_new");
		return;
	}

	// 認識を行うインスタンスを生成
	recog_ = j_create_instance_from_jconf(jconf);
	if (recog_ == nullptr) {
		Emit("error", "Error@j_create_instance_from_jconf");
		return;
	}

	// マイク入力の初期化
	if (j_adin_init(recog_) == FALSE) {
		Emit("error", "Error@j_adin_init");
		return;
	}

	// コールバックを追加
	add_callbacks();
}


void Julius::free_recog()
{
	if (recog_ == nullptr) {
		return;
	}

	j_close_stream(recog_);

	j_jconf_free(recog_->jconf);
	recog_->jconf = nullptr;

	j_recog_free(recog_);
	recog_ = nullptr;
}


void Julius::add_callbacks()
{
	// コールバックを追加する
	// --------------------------------------------------------------------------------
	// 待機時 JavaScript 側のコールバック呼び出し
	add_callback(CALLBACK_EVENT_SPEECH_READY, [](Recog* recog, void* ptr) {
		auto _this = static_cast<Julius*>(ptr);
		_this->Emit("speechReady");
	}, this);

	// 発話開始時 JavaScript 側のコールバック呼び出し
	add_callback(CALLBACK_EVENT_SPEECH_START, [](Recog* recog, void* ptr) {
		auto _this = static_cast<Julius*>(ptr);
		_this->Emit("speechStart");
	}, this);

	// 発話終了時 JavaScript 側のコールバック呼び出し
	add_callback(CALLBACK_EVENT_SPEECH_STOP, [](Recog* recog, void* ptr) {
		auto _this = static_cast<Julius*>(ptr);
		_this->Emit("speechStop");
	}, this);

	// ポーズ時 JavaScript 側のコールバック呼び出し
	add_callback(CALLBACK_EVENT_PAUSE, [](Recog* recog, void* ptr) {
		auto _this = static_cast<Julius*>(ptr);
		_this->Emit("pause");
	}, this);

	// 結果が返された時の JavaScript 側のコールバック呼び出し
	add_callback(CALLBACK_RESULT, [](Recog* recog, void* ptr) {
		auto _this = static_cast<Julius*>(ptr);
		_this->on_result(recog);
	}, this);
}


void Julius::on_result(Recog* recog)
{
	// 結果を走査
	for (const RecogProcess *r = recog->process_list; r; r = r->next) {
		WORD_INFO *winfo = r->lm->winfo;

		// 仮説の数に応じてループ
		for (int n = 0; n < r->result.sentnum; ++n) {
			// Windows だと起こらないらしいが Ubuntu だとたまに起こる…
			if (r->result.sent == nullptr) break;

			auto s     = r->result.sent[n];
			auto seq   = s.word;
			int seqnum = s.word_num;

			// 認識結果の文章を取得
			std::string output;
			for (int i = 1; i < seqnum-1; ++i) {
				// result[n][i] = winfo->woutput[seq[i]];
				output += winfo->woutput[seq[i]];
			}

			// JavaScript のコールバックを呼ぶ
			Emit("result", output);
		}
	}
}


int Julius::add_callback(const int code, callback func, void* ptr)
{
	return callback_add(recog_, code, func, ptr);
}


bool Julius::delete_callback(const int id)
{
	return callback_delete(recog_, id);
}

/* ------------------------------------------------------------------------- */
//  node.js の addon 化
/* ------------------------------------------------------------------------- */
void init(Handle<Object> target)
{
	assert( uv_mutex_init(&m) == 0 );
	emit_symbol = NODE_PSYMBOL("emit");
	Julius::Init(target);
}

NODE_MODULE(julius, init)
